python/tvm/topi/arm_cpu/mprofile/dsp/micro_kernel/multi_channel_convolve.py

```diff
@@ -179,20 +179,26 @@ def _dual_int16_channel_convolve_impl(_tensor_h, tensor_w, channels, kernel_h, k
         extern "C"
         #endif
         int32_t {_get_func_name("int16", tensor_w, channels, kernel_h, kernel_w, suffix)}(
-            uint32_t *out,
-            uint32_t *tensor,
-            uint32_t *kernel) {{
+            int32_t *out,
+            int16_t *tensor,
+            int16_t *kernel) {{
 
-          uint32_t sum_c0 = 0;
-          uint32_t sum_c1 = 0;
+          int32_t sum_c0 = 0;
+          int32_t sum_c1 = 0;
+
+          int32_t kernel_i32[{kernel_h} * {kernel_w}];
+          memcpy(kernel_i32, kernel, {kernel_h} * {kernel_w} * sizeof(int32_t));
```

Comment:
Hi Gavin, this is a functional change to remove the type punning error.  So performance can be worse.
But, it is not right to rely on undefined behaviours to have better performance.
Do you know any other way to address this undefined behaviour, without the `memcpy`s?

Reply:
If there is no alternative way to fix type punning without memcpy, we need to update other schedules too  :smile_cat:

Reply:
Perhaps I'm not understanding - why can't we just do `static_cast`?

Reply:
@NicolaLancellotti Is it the `size` of `memcpy` correct here? It seems that `kernel` is type `int16_t`, not `int32_t`, so seems copying beyond `kernel` end?

Reply:
@guberti these schedules are compiled as C not C++ aren't they? Also, last time I checked the output of one of these type punning fixes, the memcpy was optimised into two loads which can work unaligned, rather than one load of two words which only works if properly aligned - still faster than a copy 😸

Reply:
Oops, I'm being silly here. This seems fine to me - thanks @Mousius for checking.

Reply:
@gromero I've checked and `kernel_i32` has `kernel_h * kernel_w` elements of `int32_t`, the `int16_t` kernel has twice as many elements. So the size is correct, (with a lower value it doesn't work, I checked it too).